src/index.js

```diff
@@ -676,21 +699,80 @@ export default class Bynder {
       const chunk = body.slice(start, end);
       const sha256 = create256HexHash(chunk);
 
-      await this.api.send('POST', `v7/file_cmds/upload/${fileId}/chunk/${this._chunkNumber}`, chunk, {
-        additionalHeaders: {
-          'Content-SHA256': sha256
-        }
-      })
-        .catch(error => {
-          throw new Error(`[${error.message} - ${error.status}] Chunk ${this._chunkNumber} not uploaded: ${error.body}`, error);
-        });
+      await this._uploadChunk(chunk, this._chunkNumber, fileId, sha256);
 
       this._chunkNumber++;
     }
 
     return this._chunks;
   }
 
+  /**
+   * Reads a file stream and uploads it to the FS
+   *
+   * @access private
+   * @param {Object} stream File stream
+   * @param {String} fileId File ID
+   * @returns {Promise<number>}
+   */
+  _uploadStreamFile(stream, fileId) {
+    // We need to force our chunk size on the stream
+    stream._readableState.highWaterMark = FILE_CHUNK_SIZE;
+    // Developers can use this to track file upload progress
+    this._chunkNumber = 0;
+
+    return new Promise((resolve, reject) => {
+      stream.on('data', async chunk => {
+        // Hol' up!
+        stream.pause();
+        const sha256 = create256HexHash(chunk);
+
+        await this._uploadChunk(chunk, this._chunkNumber, fileId, sha256)
+          .catch(reject);
+
+        this._chunkNumber++;
+        // Continue!
+        stream.resume();
+      });
+
+      stream.on('error', reject);
+
+      stream.on('end', () => resolve(this._chunkNumber));
+    });
+  }
+
+  /**
+   * Uploads arbirtrarily sized buffer or stream file to our file service in chunks.
+   * Resolves the passed init result and final chunk number.
+   * @async
+   * @access private
+   * @param {Object} file An object containing the id of the desired collection.
+   * @param {String} file.filename The file name of the file to be saved
+   * @param {Buffer|Readable} file.body The file to be uploaded. Can be either buffer or a read stream.
+   * @param {Number} file.length The length of the file to be uploaded
+   * @param {String} fileId Unique file identifier
+   * @param {Number} size File byte size
+   * @return {Promisee<number>} Total number of chunks uploaded to the upload payload
+   */
+  async _uploadFileInChunks(file, fileId, size, bodyType) {
+    const { body } = file;
+
+    switch (bodyType) {
+    case bodyTypes.BUFFER:
+      await this._uploadBufferFile(body, fileId, size);
+      break;
+
+    case bodyTypes.STREAM:
+      this._chunks = await this._uploadStreamFile(body, fileId);
+      break;
```

Comment:
Do we need to support both options? And wouldn't it be easier to just pass the file path, instead of a stream / buffer? This is how we do it in the other SDKs but I'm not sure about the history of this SDK.

Reply:
In Node, Buffers are usually used for small size files. Streams are commonly used for larger ones.

Although we could ask for the path, I would prefer to leave this decision on the developer hands. That way we could avoid having the calculate file size each time.

Reply:
On this SDK in particular, we have always expected a Buffer or Stream file body. Not a path.